test/find_j_moves_test.cpp

```diff
@@ -0,0 +1,166 @@
+#include <map>
+#include <algorithm>
+#include <initializer_list>
+#include <gtest/gtest.h>
+#include <wl/game_state.h>
+#include <wl/find_j_moves.h>
+#include <wl/default_map.h>
+
+namespace wl {
+std::ostream& operator<<(std::ostream& os, JackResource resource) {
+	switch (resource) {
+	case JackResource::CARRIAGE: os << "CARRIAGE"; break;
+	case JackResource::ALLEY: os << "ALLEY"; break;
+	case JackResource::BOAT: os << "BOAT"; break;
+	}
+	return os;
+}
+
+std::ostream& operator<<(std::ostream& os, JackMove move) {
+	os << "JackMove{destination:" << static_cast<size_t>(move.destination());
+	if (const auto cost = move.cost(); cost.has_value()) {
+		os << ",cost:" << *cost;
+	}
+	os << "}";
+	return os;
+}
+} // namespace wl
+
+namespace {
+const wl::MapGraph k_default_map = wl::default_map();
+
+
+void do_normal_moves_test(std::initializer_list<wl::JackMove> expected_moves,
+	                      const wl::GameState& game_state,
+	                      const wl::History& history)
+{
+	// setup a map to keep track of which expected moves we found
+	std::map<wl::JackMove, bool> expected_moves_found;
+	for (const wl::JackMove expected_move : expected_moves) {
+		expected_moves_found.emplace(expected_move, false);
+	}
+	ASSERT_EQ(expected_moves_found.size(), expected_moves.size());
+
+	// find the moves
+	const auto found_moves = available_normal_jack_moves(game_state, history, k_default_map);
+	EXPECT_EQ(expected_moves_found.size(), found_moves.size());
+
+	// check that all moves found were expected
+	for (const wl::JackMove move : found_moves) {
+		auto itr = expected_moves_found.find(move);
+		if (itr == expected_moves_found.end()) {
+			FAIL() << "found unexpected move: " << move;
+			continue;
+		}
+		// note that we found this move
+		itr->second = true;
+	}
+
+	// check that all expected moves were found
+	for (const auto& keyval : expected_moves_found) {
+		if (!keyval.second) {
+			FAIL() << "did not find expected move: " << keyval.first;
+		}
+	}
```

Comment:
Go through all of the {expected_move, found_it} associated entries in the map. Make sure that each of them was actually found.

The combination of:
 - the number of things found was the number of things expected to be found
 - checking that all things found were expected to be found
 - all things expected to be found were found at least once
can be logically extended to mean that each expected thing was found exactly once

Reply:
Btw, the `auto` in `const auto& keyval` is `std::pair<wl::JackMove,bool>` *(the type of each entry in the map)*.

Thus, `!keyval.second` is referring to the `bool` member of that pair, ie. whether we found that move.